exploration/number-selection.md

```diff
@@ -92,44 +95,303 @@ ICU MF1 messages using `plural` and `selectordinal` should be representable in M
+- `plural` selection matches the operand to explicit numeric keys exactly
+  or to plural rule categories if there is no explicit match
+- `ordinal` selection matches the operand to explicit numeric keys exactly
+  or to ordinal rule categories if there is no explicit match
+
+
+### Functions
+
+The following functions use numeric selection:
+
+The function `:number` is the default selector for numeric values.
+
+The function `:plural` operates identically to `:number`, except:
+- the `select` type of this selector is always `plural`
+- the function is **not** valid as a formatter
+
+The function `:ordinal` operates identically to `:number`, except:
+- the `select` type of this selector is always `ordinal`
+- formatting behavior of this function is implementation defined
+
+> [!NOTE]
+> CLDR data has known shortcomings related to automatic formatting
+> of numbers as ordinals.
+> The default registry does not require implementations to support
+> formatting of numeric values as ordinals.
+> Users are cautioned to avoid using `:ordinal` for formatting in this release
+> in cases where message portability is a concern,
+> as many implementations will emit an error.
+
+The function `:integer` operates identically to `:number`, except:
+- the `minimumFractionDigits` and `maximumFractionDigits` options
+  default to `0` for all values
+- the _options_ `minimumFractionDigits` and `maximumFractionDigits`
+  are invalid.
 
-The default `plural` value is presumed to be the most common use case,
-and it affords the least bad fallback when used incorrectly:
-Using "plural" for "exact" still selects exactly matching cases,
-whereas using "exact" for "plural" will not select LDML category matches.
-This might not be noticeable in the source language,
+### Operands
+
+The _operand_ of a number function is either an implementation-defined numeric type or
+a literal that matches the `number-literal` production in the [ABNF](/main/spec/message.abnf).
+All other values produce a _Selection Error_ when evaluated for selection
+or a _Formatting Error_ when attempting to format the value.
+
+> For example, in Java, any subclass of `java.lang.Number` plus the primitive
+> types (`byte`, `short`, `int`, `long`, `float`, `double`, etc.) 
+> might be considered as the "implementation-defined numeric types".
+> Implementations in other programming languages would define different types
+> or classes according to their local needs.
+
+> [!NOTE]
+> String values passed as variables in the _formatting context_'s
+> _input mapping_ can be formatted as numeric values as long as their
+> contents match the `number-literal` production in the [ABNF](/main/spec/message.abnf).
+>
+> For example, if the value of the variable `num` were the string
+> `-1234.567`, it would behave identically to the local
+> variable in this example:
+> ```
+> .local $example = {|-1234.567| :number}
+> {{{$num :number} == {$example}}}
+> ```
+
+### Options
```

Comment:
Issue: I really don't think this matrix of options belongs in the MF spec.

Number formatting options are very complex. I've spent a lot of time over the last several years trying to reduce bugs and align more with client needs.

There are two parts of the model for number formatting that I think should land in MF2:

1. Specific functions aligned to specific use cases. Rather than an all-in-one `:number` function, instead tailor them to specific use cases. Engines can implement the functions to serve their use case, not necessarily being bound to a specific options matrix.
    - If we _must_ include number formatting options in MF2, we need to open a more deep discussion about which options to include and how they should behave.
2. The numeric type should be a Decimal. I think this is consistent with the `number-literal` grammar except that the extracted number should be required to be taken at face value, including trailing zeros. In other words, the "implementation-defined numeric type" must have capabilities similar to `java.math.BigDecimal`.
    - This eliminates the need for the `number` function to include options such as `minimumFractionDigits`. It also fixes one long-standing bug in MF1 which is that the `plural` function _needs to know_ the final rounded number and all of its digits in order to return the correct plural form. By pushing this out of MessageFormat, MF2 can be simpler and more accurate.

Reply:
I think what you're suggesting is a different approach than what MFWG has been working on. The syntax of a message necessarily includes the options needed for users (developers, translators) to control which patterns are selected for formatting and, relatedly, how values are inserted or formatted into those patterns.

If we didn't provide the options, each implementation on each platform would be different. Users would have to relearn the details of number and date formatting for each implementation. Tools (i.e. CAT tools) would have to know about each variation and which variation to apply to each message. This is antithetical to our goals.

> The numeric type should be a Decimal

MFv2 does not have types. We require certain formats of string literal to be accepted as inputs for some functions (number and datetime are examples) and we allow implementation-specific types to be formatted. But MFv2 does not itself have types. 

The `number-literal` in our syntax is misleading. It exists **_solely_** to allow string literals that represent numbers to be unquoted in the syntax. Otherwise numbers like `42` and `-57.3` would have to be quoted in expressions and selector keys using our "beloved" quoting syntax (note that quoting **_is_** _permitted_). See examples below. Note that number literals and "word token" literals are the most common literals in current messages.

```
.local $unquoted = {42 :number}
.local $quoted   = {|42| :number}

.match {$num}
|42|  {{Quoted}}
42    {{Unquoted}}
|-42| {{Quoted}}
-42   {{Unquoted}}
```

> This eliminates the need for the number function to include options such as minimumFractionDigits

The options in the spec are for *formatting*, not for the number itself. They provide control over the kinds of display features that UX designers want developers to implement. How else do users get a number to show a specific number of fraction digits? Or zero fill a number?

> It also fixes one long-standing bug in MF1 which is that the plural function needs to know the final rounded number and all of its digits in order to return the correct plural form. By pushing this out of MessageFormat, MF2 can be simpler and more accurate.

There are two ways to look at this.

On the one hand, it is not a "bug" to know that the plural selector for `2` should potentially select a different category from the plural selector for `2.0`.

On the other hand, read this design doc carefully: **MessageFormat** does not know anything about how plural selection work, how numbers are formatted, what the correct seperators are, usw. 

What MF2 requires is that implementers provide the minimum set of functions we specify and that those functions accept certain options. We value message portability, not because we think messages will be formatted across many implementations but because portability guarantees that what you learn about how to format a message is transferable (and note that "what you learn" means "what your tools can assume" too).

> If we must include number formatting options in MF2, we need to open a more deep discussion about which options to include and how they should behave.

I think we absolutely must format numbers and date/time values. No one needs MF2 if all we do is insert strings into strings. This design doc *is* that discussion. This is on the [F2F agenda](https://docs.google.com/document/d/1NSeimrSFzuKxne7BHmuo4Nseusa1Gfoif0fBdbF2obU/edit) for tomorrow morning, albeit not the first item in the agenda.

If you want to make an alternate proposal, we *require* a design doc--I would absolutely welcome a constructive proposal. You can make a PR against this design doc, perhaps adding an alternative considered or you can make a new one. The template is under "exploration".

Reply:
> I think what you're suggesting is a different approach than what MFWG has been working on. The syntax of a message necessarily includes the options needed for users (developers, translators) to control which patterns are selected for formatting and, relatedly, how values are inserted or formatted into those patterns. ... If we didn't provide the options, each implementation on each platform would be different. Users would have to relearn the details of number and date formatting for each implementation. Tools (i.e. CAT tools) would have to know about each variation and which variation to apply to each message. This is antithetical to our goals.

I think I'm aligned on MF2 taking the role of a common set of operations to align implementations across the ecosystem. I'm just not aligned on a giant options bag as necessarily being the right mechanism to do that. I was suggesting that we take a more use-case-driven approach.

> MFv2 does not have types. We require certain formats of string literal to be accepted as inputs for some functions (number and datetime are examples) and we allow implementation-specific types to be formatted. But MFv2 does not itself have types. ... The options in the spec are for formatting, not for the number itself. They provide control over the kinds of display features that UX designers want developers to implement. How else do users get a number to show a specific number of fraction digits? Or zero fill a number?

I'm trying to make the point that specifying a number of fraction digits _should not_ be a formatting option. There is a difference between "1" and "1.0" (the latter has greater precision); the trailing zeros should be part of the data model. MF2 should request that implementations prefer numeric types that represent a formattable decimal number.

> The number-literal in our syntax is misleading. It exists solely to allow string literals that represent numbers to be unquoted in the syntax.

I was referring to this paragraph in the PR: "The _operand_ of a number function is either an implementation-defined numeric type or a literal that matches the `number-literal` production in the [ABNF](/main/spec/message.abnf)."

I think "implementation-defined numeric type" should be a type that is capable of representing trailing zeros. Failing that, a string matching the `number-literal` production is acceptable as input to the `:number` function.

> I think we absolutely must format numbers and date/time values.

I have similarly well formed opinions on date/time formatting options and would like to weigh in on that discussion as well.

Reply:
> I think "implementation-defined numeric type" should be a type that is capable of representing trailing zeros. Failing that, a string matching the number-literal production is acceptable as input to the :number function.

Note that "operand" in our spec means "input". Consider this pseudo-code fragment:

```java
MessageFormat2 mf2 = MessageFormat2.withPattern("The number is {$num}");
ImmutableMap args = ImmutableMap.of("num", aNumber);
System.out.println(mf2.format(args).toString());
```

In this example, `$num` is the _operand_. And its value is whatever `aNumber` is. The text you quote says the value `aNumber` can be whatever `MessageFormat2` accepts. This could be an `int` or a `double`. It could be a `BigDecimal`. *Any* numeric type might reasonably be input. As a user, my expectation is that I don't have to do anything to my numbers in order to format them.

I think what you're talking about is the *output*. In ICU, the result of the number formatter is likely to be a `FormattedNumber` whose internal representation is probably a `BigDecimal`. 

MF2 doesn't care.

Reply:
If the input is an int or a float, that's fine. You just display the numeric value. And if you want trailing zeros, you should use BigDecimal as your input. We're just moving the concept of trailing zeros from a formatting concern to a data concern, which is where I believe it belongs.